ipam/ring/ring.go

```diff
@@ -368,31 +372,34 @@ func (r *Ring) AllRangeInfo() (result []RangeInfo) {
 }
 
 // ClaimForPeers claims the entire ring for the array of peers passed
-// in.  Only works for empty rings.
+// in.  Only works for empty rings. Each claimed range is CIDR-aligned.
 func (r *Ring) ClaimForPeers(peers []mesh.PeerName) {
 	common.Assert(r.Empty())
+
+	defer r.callUpdateCallback()()
 	defer r.assertInvariants()
 	defer r.updateExportedVariables()
-
-	totalSize := r.distance(r.Start, r.End)
-	share := totalSize/address.Count(len(peers)) + 1
-	remainder := totalSize % address.Count(len(peers))
-	pos := r.Start
-
-	for i, peer := range peers {
-		if address.Count(i) == remainder {
-			share--
-			if share == 0 {
-				break
-			}
+	defer func() {
+		e := r.Entries[len(r.Entries)-1]
+		common.Assert(address.Add(e.Token, address.Offset(e.Free)) == r.End)
+	}()
+
+	var subdivide func(from, to address.Address, peers []mesh.PeerName)
+	subdivide = func(from, to address.Address, peers []mesh.PeerName) {
+		share := address.Length(to, from)
+		if share == 0 {
+			return
 		}
-
-		r.Entries.insert(entry{Token: pos, Peer: peer, Free: share})
-		pos += address.Address(share)
+		if share == 1 || len(peers) == 1 {
+			r.Entries.insert(entry{Token: from, Peer: peers[0], Free: share})
+			return
+		}
+		mid := address.Add(from, address.Offset(share/2))
+		subdivide(from, mid, peers[:len(peers)/2])
+		subdivide(address.Add(mid, address.Offset(share%2)), to, peers[len(peers)/2:])
```

Comment:
eek. I don't think @bboreham or I realised that we'd need a closure here. So I think this ought to be a separate function after all. Sorry.